README.md
# Bibata Zebra

'Semi-Animated' Bibata cursors with zebra stripes.

[![build](https://github.com/ful1e5/Bibata-Zebra-Cursor/actions/workflows/build.yml/badge.svg)](https://github.com/ful1e5/Bibata-Zebra-Cursor/actions/workflows/build.yml)
[![CodeFactor](https://www.codefactor.io/repository/github/ful1e5/bibata-zebra-cursor/badge)](https://www.codefactor.io/repository/github/ful1e5/bibata-zebra-cursor)
[![Twitter](https://img.shields.io/badge/twitter-ful1e5-blue)](https://twitter.com/ful1e5)

### Styles

- **Original :** Sharp edge Bibata Cursors
- **Modern :** Round edge Bibata Cursors

### Cursor Sizes

<kbd>22</kbd>
<kbd>24</kbd>
<kbd>28</kbd>
<kbd>32</kbd>
<kbd>40</kbd>
<kbd>48</kbd>
<kbd>56</kbd>
<kbd>64</kbd>
<kbd>72</kbd>
<kbd>80</kbd>
<kbd>88</kbd>
<kbd>96</kbd>

## Quick install

<p align="center">
  <a href="https://www.pling.com/p/1570301/" >
    <img title="Bibata Pling Store" width="40%" src="https://imgur.com/VxSgrWw.png">
  </a>
</p>

## Preview

<p align="center">
  <img title="Bibata Zebra" width="90%" src="./branding/preview.svg">
  </br>
  <sub>Bibata Zebra</sub>
</p>

## Manual Install

Latest `Stable` & `Development` releases can be downloaded from [Here](https://github.com/ful1e5/Bibata-Zebra-Cursor/releases)

#### Linux/X11

```bash
# extract `Bibata-Zebra.tar.gz`
tar -xvf Bibata-Zebra.tar.gz

# For local users
mv Bibata-Zebra-* ~/.icons/

# For all users
sudo mv Bibata-Zebra-* /usr/share/icons/
```

#### Windows

1. unzip `.zip` file
2. Open unziped directory in Explorer, and **right click** on `install.inf`.
3. Click 'Install' from the context menu, and authorize the modifications to your system.
4. Open _Control Panel_ > _Personalization and Appearance_ > _Change mouse pointers_, and select **Bibata Zebra Cursors**.
5. Click '**Apply**'.

### Uninstall

#### Linux/X11

```bash
# For local users
rm ~/.icons/Bibata-Zebra-*

# For all users
sudo rm /usr/share/icons/Bibata-Zebra-*
```

#### Windows

1. Go to **Registry Editor** by typing the same in the _start search box_.
2. Expand `HKEY_CURRENT_USER` folder and expand `Control Panel` folder.
3. Go to `Cursors` folder and click on `Schemes` folder - all the available custom cursors that are installed will be listed here.
4. **Right Click** on the name of cursor file you want to uninstall; for eg.: _Bibata-Zebra Cursors_ and click `Delete`.
5. Click '**yes**' when prompted.

# Dependencies

## External Libraries

- libxcursor-dev
- libx11-dev
- libpng-dev (<=1.6)

#### Install External Libraries

##### macOS

```bash
brew install --cask xquartz
brew install libpng
```

##### Debain/ubuntu

```bash
sudo apt install libx11-dev libxcursor-dev libpng-dev
```

##### ArchLinux/Manjaro

```bash
sudo pacman -S libx11 libxcursor libpng
```

##### Fedora/Fedora Silverblue/CentOS/RHEL

```bash
sudo dnf install libX11-devel libXcursor-devel libpng-devel
```

## Build Dependencies

- [gcc](https://gcc.gnu.org/install/)
- [make](https://www.gnu.org/software/make/)
- [nodejs](https://nodejs.org/en/) (<=12.x.x)
- [yarn](https://classic.yarnpkg.com/en/docs/install/)
- [python](https://www.python.org/downloads/) (<=3.8)
- [pip3](https://pip.pypa.io/en/stable/installing/)

### Node Packages

- [puppeteer](https://www.npmjs.com/package/puppeteer)
- [pngjs](https://www.npmjs.com/package/pngjs)
- [pixelmatch](https://www.npmjs.com/package/pixelmatch)

### PyPi Packages

- [clickgen](https://pypi.org/project/clickgen/s)

## Build From Scratch

### ⚡ Auto Build (using GitHub Actions)

GitHub Actions is automatically runs on every `push`(on **main** & **dev** branch) and `pull request`(on **main** branch), You found theme resources in `artifact` section of **ci**. GitHub **Actions** available inside [.github/workflows](https://github.com/ful1e5/Bibata-Zebra-Cursor)/tree/main/.github/workflows) directory.

### Manual Build

> Check **[Makefile](./Makefile)** for more targets.

```bash
make
```

#### Build Only `XCursor` theme

```bash
make unix
```

#### Customize `XCursor` size

```bash
make unix X_SIZES=22            # Only built '22px' pixel-size.
make unix X_SIZES=22 24 32      # Multiple sizes are provided with  ' '(Space)
```

#### Install `XCursor` theme

```bash
make install            # install as user
  # OR
sudo make install       # install as root
```

#### Build Only `Windows` theme

```bash
make windows
```

#### Customize `Windows Cursor` size

```bash
make windows WIN_SIZE=96            # Supports only one pixel-size
```

> Windows installations steps are same as [these](#windows).

# You may also like...

- [**Bibata**](https://github.com/ful1e5/Bibata_Cursor) - Material based cursors
- [**Bibata Extra**](https://github.com/ful1e5/Bibata_Extra_Cursor) - More Bibata!
- [**Bibata Rainbow**](https://github.com/ful1e5/Bibata_Cursor_Rainbow)
- [**Bibata Adapta**](https://gitlab.com/cscs/Bibata_AdaptaBreath_Cursors) - Bibata Based Cursor Made for AdaptaBreath and Manjaro.
- [**Bibata Translucent**](https://github.com/Silicasandwhich/Bibata_Cursor_Translucent) - Bibata translucent is a translucent flavor of the Bibata.

# Bugs

Bugs should be reported [here](https://github.com/ful1e5/Bibata-Zebra-Cursor/issues) on the Github issues page.

# Getting Help

You can create a **issue**, I will help you. 🙂

# Contributing

Check [CONTRIBUTING.md](CONTRIBUTING.md), any suggestions for features and contributions to the continuing code masterelopment can be made via the issue tracker or code contributions via a `Fork` & `Pull requests`.

# Credit

- [Bibata](https://github.com/ful1e5/Bibata_Cursor)
- [Bibata Rainbow](https://github.com/ful1e5/Bibata_Cursor_Rainbow)
- [Adwaita](https://github.com/GNOME/adwaita-icon-theme)
- [Dmz](https://github.com/GalliumOS/dmz-cursor-theme)
- [Yaru](https://github.com/ubuntu/yaru)
- Emojis are taken from [here](https://emojipedia.org/)
- Wedge loading from [loading.io](https://loading.io/spinner/wedges/-pie-wedge-pizza-circle-round-rotate) with **Microsoft** colors


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [unreleased]

### Added

- support creator with liberapay
- funding file ignored in `build` GitHub Action
- Add symlink for `top_left_arrow` cursor ful1e5/BreezeX_Cursor#10 ful1e5/BreezeX_Cursor#11
- Uninstall docs ful1e5/apple_cursor#79 ful1e5/apple_cursor#80

### Changed

- GitHub action renamed to `build`
- Minimal README.md
- emojis removed from README.md
- Fix `None` value warning in `builder/symlinks.py`
- sponsorships updated

## [v1.0.0] - 7 Aug 2021

### Added

- Initial release 🎊
- Logo and badges
- CI/CD Pipelines

[unreleased]: https://github.com/ful1e5/Bibata-Zebra-Cursor/compare/v1.0.0...main
[v1.0.0]: https://github.com/ful1e5/Bibata-Zebra-Cursor/tree/v1.0.0


builder/build.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
from pathlib import Path

from src.configure import get_config
from src.generator import Info, build, wbuild, xbuild

parser = argparse.ArgumentParser(
    prog="bibata_zebra_builder",
    description="'Bibata Zebra' cursor build python script.",
)

# Positional Args.
parser.add_argument(
    "platform",
    choices=("windows", "unix", "all"),
    default="all",
    const="all",
    nargs="?",
    help="Set package type, Which you want to build. (default: '%(default)s')",
)

# Optional Args.
parser.add_argument(
    "-p",
    "--png-dir",
    dest="png_dir",
    metavar="PNG",
    type=str,
    default="../bitmaps",
    help="To change pngs directory. (default: %(default)s)",
)

parser.add_argument(
    "-o",
    "--out-dir",
    dest="out_dir",
    metavar="OUT",
    type=str,
    default="../themes",
    help="To change output directory. (default: %(default)s)",
)


parser.add_argument(
    "-xs",
    "--xsizes",
    dest="xsizes",
    metavar="INT",
    nargs="+",
    default=[
        22,
        24,
        28,
        32,
        40,
        48,
        56,
        64,
        72,
        80,
        88,
        96,
    ],
    type=int,
    help="Set pixel-size for xcursor. (default: %(default)s)",
)


parser.add_argument(
    "-ws",
    "--win-size",
    dest="win_size",
    metavar="INT",
    default=24,
    type=int,
    help="Set pixel-size for Windows cursors. (default: %(default)s)",
)


parser.add_argument(
    "-wcs",
    "--win-canvas-size",
    dest="win_canvas_size",
    metavar="INT",
    default=32,
    type=int,
    help="Set pixel-size for Windows cursor's canvas. (default: %(default)s)",
)

# Preparing build
args = parser.parse_args()

bitmaps_dir = Path(args.png_dir).absolute()
name = bitmaps_dir.stem
comments = {
    "Bibata-Zebra-Modern": "Rounded edge Bibata cursors with Zebra stripes",
    "Bibata-Zebra-Original": "Sharp edge Bibata cursors with Zebra stripes",
}


x_out_dir = Path(args.out_dir) / name
win_out_dir = Path(args.out_dir) / f"{name}-Windows"

print(f"Getting '{name}' bitmaps ready for build...")

config = get_config(
    bitmaps_dir,
    x_sizes=args.xsizes,
    win_canvas_size=args.win_canvas_size,
    win_size=args.win_size,
)

info = Info(name=name, comment=comments.get(name, f"{name} Cursors"))

if args.platform == "unix":
    xbuild(config, x_out_dir, info)
elif args.platform == "windows":
    wbuild(config, win_out_dir, info)
else:
    build(config, x_out_dir, win_out_dir, info)


builder/src/configure.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from pathlib import Path
from typing import Any, Dict, Tuple, TypeVar, Union

from clickgen.util import PNGProvider

from .constants import WIN_CURSORS_CFG, WIN_DELAY, X_CURSORS_CFG, X_DELAY

X = TypeVar("X")


def to_tuple(x: X) -> Tuple[X, X]:
    return (x, x)


def get_config(bitmaps_dir: Union[str, Path], **kwargs) -> Dict[str, Any]:
    """Return configuration of `Bibata Zebra` pointers.

    :param bitmaps_dir: Path to .png file's directory.
    :type bitmaps_dir: Union[str, Path]

    :param x_sizes: List of pixel-sizes for xcursors.
    :type x_sizes: List[int]

    :param win_canvas_size: Windows cursor's canvas pixel-size.
    :type win_canvas_size: int

    :param win_size: Pixel-size for Windows cursor.
    :type win_size: int

    Example:

    ```python
        get_config(
            "./bitmaps",
            x_sizes=[(24, 24), (32, 32)],
            win_canvas_size=(32, 32),
            win_size=(24, 24),
        )
    ```
    """

    w_size = to_tuple(kwargs.pop("win_size"))
    w_canvas_size = to_tuple(kwargs.pop("win_canvas_size"))
    raw_x_sizes = kwargs.pop("x_sizes")
    x_sizes = [to_tuple(size) for size in raw_x_sizes]
    config: Dict[str, Any] = {}

    for key, item in X_CURSORS_CFG.items():
        x_hot: int = int(item.get("xhot", 0))
        y_hot: int = int(item.get("yhot", 0))
        hotspot: Tuple[int, int] = (x_hot, y_hot)

        delay: int = int(item.get("delay", X_DELAY))
        pngs = PNGProvider(bitmaps_dir).get(key)

        if not pngs:
            raise FileNotFoundError(f"{key} not found in {bitmaps_dir}")

        data = {
            "png": pngs,
            "x_sizes": x_sizes,
            "hotspot": hotspot,
            "delay": delay,
        }

        win_data = WIN_CURSORS_CFG.get(key)

        if win_data:
            win_key: str = str(win_data.get("to"))

            position: str = str(win_data.get("position", "center"))
            win_delay: int = int(win_data.get("delay", WIN_DELAY))

            canvas_size = win_data.get("canvas_size", w_canvas_size)
            win_size = win_data.get("size", w_size)

            # Because provided cursor size is bigger than cursor's canvas.
            # Also, "position" settings will not effect on cursor because the
            # cursor's canvas and cursor sizes are equals.
            if (win_size[0] > canvas_size[0]) | (win_size[1] > canvas_size[1]):
                canvas_size = win_size

            config[key] = {
                **data,
                "win_key": win_key,
                "position": position,
                "canvas_size": canvas_size,
                "win_size": win_size,
                "win_delay": win_delay,
            }
        else:
            config[key] = data

    return config


builder/src/generator.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from pathlib import Path
from typing import Any, Dict, NamedTuple

from clickgen.builders import WindowsCursor, XCursor
from clickgen.core import CursorAlias
from clickgen.packagers import WindowsPackager, XPackager

from .constants import AUTHOR, URL
from .symlinks import add_missing_xcursor


class Info(NamedTuple):
    """Important cursor package's information field"""

    name: str
    comment: str


def xbuild(config: Dict[str, Dict[str, Any]], x_out_dir: Path, info: Info) -> None:
    """Build `Bibata Zebra` cursor theme for only `X11`(UNIX) platform.

    :param config: `Bibata` configuration.
    :type config: Dict[str, Dict[str, Any]]

    :param x_out_dir: Path to the output directory, \
                      Where the `X11` cursor theme package will generate.\
                      It also creates a directory if not exists.
    :type x_out_dir: Path

    :param info: Content theme name & comment
    :type info: Info
    """

    for _, item in config.items():
        png = item["png"]
        hotspot = item["hotspot"]
        x_sizes = item["x_sizes"]
        delay = item["delay"]

        with CursorAlias.from_bitmap(png, hotspot) as alias:
            x_cfg = alias.create(x_sizes, delay)
            print(f"Building '{x_cfg.stem}' XCursor...")
            XCursor.create(x_cfg, x_out_dir)

    add_missing_xcursor(x_out_dir / "cursors")
    XPackager(x_out_dir, info.name, info.comment)


def wbuild(config: Dict[str, Dict[str, Any]], win_out_dir: Path, info: Info) -> None:
    """Build `Bibata Zebra` cursor theme for only `Windows` platforms.

    :param config: `Bibata` configuration.
    :type config: Dict[str, Dict[str, Any]]

    :param win_out_dir: Path to the output directory,\
                        Where the `Windows` cursor theme package will \
                        generate. It also creates a directory if not exists.
    :type win_out_dir: Path

    :param info: Content theme name & comment
    :type info: Dict
    """

    for _, item in config.items():
        png = item["png"]
        hotspot = item["hotspot"]
        x_sizes = item["x_sizes"]
        delay = item["delay"]

        with CursorAlias.from_bitmap(png, hotspot) as alias:
            alias.create(x_sizes, delay)

            if item.get("win_key"):
                position = item["position"]
                win_size = item["win_size"]
                win_key = item["win_key"]
                canvas_size = item["canvas_size"]
                win_delay = item["win_delay"]

                win_cfg = alias.reproduce(
                    win_size, canvas_size, position, delay=win_delay
                ).rename(win_key)
                print(f"Building '{win_cfg.stem}' Windows Cursor...")
                WindowsCursor.create(win_cfg, win_out_dir)

    WindowsPackager(win_out_dir, info.name, info.comment, AUTHOR, URL)


def build(
    config: Dict[str, Dict[str, Any]], x_out_dir: Path, win_out_dir: Path, info: Info
) -> None:
    """Build `Bibata Zebra` cursor theme for `X11` & `Windows` platforms.

    :param config: `Bibata` configuration.
    :type config: Dict[str, Dict[str, Any]]

    :param win_out_dir: Path to the output directory,\
                        Where the `Windows` cursor theme package will \
                        generate. It also creates a directory if not exists.
    :type win_out_dir: Path

    :param info: Content theme name & comment
    :type info: Dict
    """

    def win_build(item: Dict[str, Any], alias: CursorAlias) -> None:
        position = item["position"]
        win_size = item["win_size"]
        win_key = item["win_key"]
        canvas_size = item["canvas_size"]
        win_delay = item["win_delay"]

        win_cfg = alias.reproduce(
            win_size, canvas_size, position, delay=win_delay
        ).rename(win_key)
        print(f"Building '{win_cfg.stem}' Windows Cursor...")
        WindowsCursor.create(win_cfg, win_out_dir)

    for _, item in config.items():
        png = item["png"]
        hotspot = item["hotspot"]
        x_sizes = item["x_sizes"]
        delay = item["delay"]

        with CursorAlias.from_bitmap(png, hotspot) as alias:
            x_cfg = alias.create(x_sizes, delay)
            print(f"Building '{x_cfg.stem}' XCursor...")
            XCursor.create(x_cfg, x_out_dir)

            if item.get("win_key"):
                win_build(item, alias)

    add_missing_xcursor(x_out_dir / "cursors")
    XPackager(x_out_dir, info.name, info.comment)

    WindowsPackager(win_out_dir, info.name, info.comment, AUTHOR, URL)


builder/src/constants.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from typing import Dict

# Info
AUTHOR = "Kaiz Khatri"
URL = "https://github.com/ful1e5/Bibata-Zebra-Cursor"

# XCursor
X_DELAY: int = 50


# Windows Cursor
WIN_DELAY = 2

X_CURSORS_CFG: Dict[str, Dict[str, int]] = {
    ##########
    # Static #
    ##########
    "crossed_circle.png": {"xhot": 100, "yhot": 100},
    "wayland-cursor.png": {"xhot": 100, "yhot": 100},
    "X_cursor.png": {"xhot": 100, "yhot": 100},
    ############
    # Animated #
    ############
    # Note: Animated cursors don't need an extension and frame numbers.
    "bd_double_arrow": {"xhot": 98, "yhot": 100},
    "bottom_left_corner": {"xhot": 31, "yhot": 172},
    "bottom_right_corner": {"xhot": 170, "yhot": 172},
    "bottom_side": {"xhot": 100, "yhot": 164},
    "bottom_tee": {"xhot": 100, "yhot": 164},
    "center_ptr": {"xhot": 98, "yhot": 131},
    "circle": {"xhot": 48, "yhot": 25},
    "context-menu": {"xhot": 48, "yhot": 25},
    "copy": {"xhot": 48, "yhot": 25},
    "cross": {"xhot": 98, "yhot": 96},
    # "crosshair": {"xhot": 99, "yhot": 99},
    "dnd_no_drop": {"xhot": 86, "yhot": 79},
    "dnd-ask": {"xhot": 86, "yhot": 79},
    "dnd-copy": {"xhot": 86, "yhot": 79},
    "dnd-link": {"xhot": 86, "yhot": 79},
    "dnd-move": {"xhot": 86, "yhot": 79},
    "dotbox": {"xhot": 100, "yhot": 100},
    "fd_double_arrow": {"xhot": 98, "yhot": 100},
    "grabbing": {"xhot": 106, "yhot": 79},
    "hand1": {"xhot": 113, "yhot": 95},
    "hand2": {"xhot": 88, "yhot": 32},
    "left_ptr": {"xhot": 53, "yhot": 36},
    "left_side": {"xhot": 35, "yhot": 100},
    "left_tee": {"xhot": 165, "yhot": 95},
    "link": {"xhot": 48, "yhot": 25},
    "ll_angle": {"xhot": 34, "yhot": 165},
    "lr_angle": {"xhot": 167, "yhot": 164},
    "move": {"xhot": 100, "yhot": 100},
    "pencil": {"xhot": 37, "yhot": 161},
    "plus": {"xhot": 100, "yhot": 100},
    "pointer-move": {"xhot": 48, "yhot": 25},
    "question_arrow": {"xhot": 102, "yhot": 102},
    "right_ptr": {"xhot": 150, "yhot": 29},
    "right_side": {"xhot": 163, "yhot": 98},
    "right_tee": {"xhot": 30, "yhot": 96},
    "sb_down_arrow": {"xhot": 100, "yhot": 126},
    "sb_h_double_arrow": {"xhot": 100, "yhot": 100},
    "sb_left_arrow": {"xhot": 86, "yhot": 100},
    "sb_right_arrow": {"xhot": 113, "yhot": 100},
    "sb_up_arrow": {"xhot": 99, "yhot": 86},
    "sb_v_double_arrow": {"xhot": 100, "yhot": 100},
    "tcross": {"xhot": 98, "yhot": 100},
    "top_left_corner": {"xhot": 29, "yhot": 27},
    "top_right_corner": {"xhot": 170, "yhot": 28},
    "top_side": {"xhot": 98, "yhot": 34},
    "top_tee": {"xhot": 98, "yhot": 29},
    "ul_angle": {"xhot": 34, "yhot": 35},
    "ur_angle": {"xhot": 164, "yhot": 34},
    "vertical-text": {"xhot": 100, "yhot": 100},
    "xterm": {"xhot": 100, "yhot": 100},
    "zoom-in": {"xhot": 90, "yhot": 89},
    "zoom-out": {"xhot": 93, "yhot": 90},
    # Truly animated
    "left_ptr_watch": {"xhot": 50, "yhot": 28, "delay": 25},
    "wait": {"xhot": 100, "yhot": 100, "delay": 25},
}

WIN_CURSORS_CFG: Dict[str, Dict[str, str]] = {
    "right_ptr": {"to": "Alternate", "position": "top_left"},
    "cross": {"to": "Cross"},
    "left_ptr": {"to": "Default", "position": "top_left"},
    "bd_double_arrow": {"to": "Diagonal_1"},
    "fd_double_arrow": {"to": "Diagonal_2"},
    "pencil": {"to": "Handwriting"},
    "question_arrow": {"to": "Help", "position": "top_left"},
    "sb_h_double_arrow": {"to": "Horizontal"},
    "xterm": {"to": "IBeam", "position": "top_left"},
    "hand2": {"to": "Link", "position": "top_left"},
    "hand1": {"to": "Move"},
    "circle": {"to": "Unavailiable", "position": "top_left"},
    "sb_v_double_arrow": {"to": "Vertical"},
    "left_ptr_watch": {"to": "Work", "position": "top_left"},
    "wait": {"to": "Busy"},
}
